### DISCUSSION TITLE
How to exclude certain packages

### DISCUSSION THREAD

**Discussion by TomBeckett** [spring-projects/spring-modulith#416] [Q&A]
Hi all! 👋

We're using JOOQ in our project to generate sources as a package for example `com.example.db`. We love how JOOQ lets us write typesafe SQL and don't want to drop it.

Unfortunately it also flags as a cyclic dependency when running a basic test with spring-modulith:

```java
    @Test
    void writeDocumentationSnippets() {
        final var modules = ApplicationModules.of(Application.class).verify();

        new Documenter(modules)
                .writeModulesAsPlantUml()
                .writeIndividualModulesAsPlantUml();
    }
```

Output
```java
org.springframework.modulith.core.Violations: - Cycle detected: Slice auth ->   // Our code
                Slice tables ->  // JOOQ generated tables
                Slice auth // Our code
```

Is there a way we can ignore the `com.example.db` package entirely?

**Comment by odrotbohm** [comment]
`ApplicationModules.of(…)` has an overload that takes a `DescribedPredicate<JavaClass>` to filter types to ignore. Thus, an `ApplicationModules.of(Application.class, JavaClass.Predicates.resideInAPackage("com.example.db"))` should do the trick.

**Reply by humoyun-norboboev** [reply]
We have been experiencing the same problem, but the exclusion did not help. The version we use is spring-modulith-bom:1.0.0. Any ideas on what can be wrong?

**Reply by AquaGenLive** [reply]
@humoyun-norboboev I had a similar problem. I forgot to add the wildcard for all packages within the excluded package. Maybe instead of `"com.example.db"` you should try `com.example.db..`.

The full documentation on how the matching works can be found in the [PackageMatcher](https://github.com/TNG/ArchUnit/blob/main/archunit/src/main/java/com/tngtech/archunit/core/domain/PackageMatcher.java) class.

**Reply by odrotbohm** [reply]
That's a helpful observation. Would you mind creating a ticket to ask for that to be clarified in the documentation?

**Reply by AquaGenLive** [reply]
Instead of describing that in a ticket, I thought it may be faster to open a PR for the doc: https://github.com/spring-projects/spring-modulith/pull/550

**Reply by humoyun-norboboev** [reply]
@AquaGenLive `com.example.db..` this one helped. Thanks!

**Comment by humoyun-norboboev** [comment]
@TomBeckett Could you please provide some details on how it solved the problem?

**Comment by marceldekoster-ahold** [comment]
For this solution does not work at all. Also not with the ..
I am using spring-modulith-1.4.2. 

I have some generated code from graphql and use that in a module. I want the graphql package to be excluded for verification entirely from all modules.

**Reply by odrotbohm** [reply]
Would you mind preparing a small reproducer and filing a ticket with that attached?

**Reply by marceldekoster-ahold** [reply]
Hmmm, I created a sample project to reproduce the issue. And there the ignore worked as described above. But in my bigger project it does not... I guess there is some more work for me to do to figure this out...

**Reply by odrotbohm** [reply]
Okay, I guess that's partially good news then. 😅 Feel free to drop a note if there's anything actionable.

**Reply by marceldekoster-ahold** [reply]
Yeah, I accidentally added `@ApplicationModuleTest` to the module verification test. Probably some vibe coding thing, but nothing helped me to figure this out, except starting from scratch with a clean springboot project with modulith, which had that tiny difference. So I removed that and all of a sudden it worked like a charm.

Thanks for the attention and I hope this mistake helps others that end up here...
